src/System.IO.Pipelines/Gate.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    /// <summary>
+    /// Simple awaitable gate - intended to synchronize a single producer with a single consumer to ensure the producer doesn't
+    /// produce until the consumer is ready. Similar to a <see cref="TaskCompletionSource{TResult}"/> but reusable so we don't have
+    /// to keep allocating new ones every time.
+    /// </summary>
+    /// <remarks>
+    /// The gate can be in one of two states: "Open", indicating that an await will immediately return and "Closed", meaning that an await
+    /// will block until the gate is opened. The gate is initially "Closed" and can be opened by a call to <see cref="Open"/>. Upon the completion
+    /// of an await, it will automatically return to the "Closed" state (this is done in the <see cref="GetResult"/> call that is injected by the
+    /// compiler's async/await logic).
+    /// </remarks>
+    internal class Gate : ICriticalNotifyCompletion
+    {
+        private static readonly Action _gateIsOpen = () => {};
+
+        private volatile Action _gateState;
+
+        /// <summary>
+        /// Returns a boolean indicating if the gate is "open"
+        /// </summary>
+        public bool IsCompleted => _gateState == _gateIsOpen;
```

Comment:
I'll file bugs for the Gate/Signal changes. I want to get this in asap since it's a giant "move the code over" PR. I'd prefer if there was a reusable TCS in the framework, these 2 types are basically that.